### PR TITLE
Fix: TDR dev dataset is stale (HumanCellAtlas/dcp2#17)

### PR DESCRIPTION
**Pull request by jessebrennan**
https://github.com/HumanCellAtlas/dcp2/issues/17

Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [ ] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [ ] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Follow upgrading instructions with `sandbox` selected
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Follow upgrading instructions with `dev` selected
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Announce that devs will need to follow upgrading instructions
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Announce whether reindexing these snapshots succeeded or failed
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: deployments/sandbox/environment.py [hunk 1]
```diff
@@ -51,26 +51,121 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-0c5c20b5:snapshot/hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907:4',
+            'tdr:datarepo-dev-e2ab8487:snapshot/hca_dev_03c6fce7789e4e78a27a664d562bb738__20210902_20210907:4',
+            'tdr:datarepo-dev-b839d6c7:snapshot/hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20210907:4',
+            'tdr:datarepo-dev-311340f6:snapshot/hca_dev_2d8460958a334f3c97d4585bafac13b4__20210902_20210907:4',
+            'tdr:datarepo-dev-4c3e6011:snapshot/hca_dev_248fcf0316c64a41b6ccaad4d894ca42__20210907_20210907:4',
+            'tdr:datarepo-dev-76de829d:snapshot/hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907:4',
+            'tdr:datarepo-dev-24e9529e:snapshot/hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908:4',
+            'tdr:datarepo-dev-87beea11:snapshot/hca_dev_ae71be1dddd84feb9bed24c3ddb6e1ad__20210916_20210916:4',
+            # Managed access snapshots:
+            'tdr:datarepo-dev-d4b988d6:snapshot/hca_dev_a004b1501c364af69bbd070c06dbc17d__20210830_20210903:4',
```
Comment (amarjandu) [inline deployments/sandbox/environment.py:166]: We might want to remove the common prefix.

Comment (amarjandu) [reply deployments/sandbox/environment.py:166]: for this snapshot, there are 16 total links, and 1 under the uuid 4.

File: deployments/sandbox/environment.py [hunk 2]
```diff
@@ -51,26 +51,121 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-02c59b72:snapshot/hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903:4',
```
Comment (amarjandu) [inline deployments/sandbox/environment.py:167]: 1190 links total, 84 under the common prefix `4`

**Comment by jessebrennan** [comment]
Since we can't attach this to its ticket via Zenhub because of a permissions issue @hannes-ucsc mentioned, I'm assigning myself so that it stays on my board.

File: deployments/sandbox/environment.py [hunk 3]
```diff
@@ -51,26 +51,121 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-24e672db:snapshot/hca_dev_cddab57b68684be4806f395ed9dd635a__20210831_20210907:4',
+            'tdr:datarepo-dev-6fdac3db:snapshot/hca_dev_e8808cc84ca0409680f2bba73600cba6__20210902_20210907:4',
+            'tdr:datarepo-dev-0c5c20b5:snapshot/hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907:4',
+            'tdr:datarepo-dev-e2ab8487:snapshot/hca_dev_03c6fce7789e4e78a27a664d562bb738__20210902_20210907:4',
+            'tdr:datarepo-dev-b839d6c7:snapshot/hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20210907:4',
+            'tdr:datarepo-dev-311340f6:snapshot/hca_dev_2d8460958a334f3c97d4585bafac13b4__20210902_20210907:4',
+            'tdr:datarepo-dev-4c3e6011:snapshot/hca_dev_248fcf0316c64a41b6ccaad4d894ca42__20210907_20210907:4',
+            'tdr:datarepo-dev-76de829d:snapshot/hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907:4',
+            'tdr:datarepo-dev-24e9529e:snapshot/hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908:4',
+            'tdr:datarepo-dev-87beea11:snapshot/hca_dev_ae71be1dddd84feb9bed24c3ddb6e1ad__20210916_20210916:4',
```
Comment (amarjandu) [inline deployments/sandbox/environment.py:164]: The project name is wrong here. `datarepo-dev-1dce87e5` should be used.
TDR team updated the spreadsheet. 
https://docs.google.com/spreadsheets/d/1yYiB3TrX0wvLCHvtWY1GTZZcAf5MmMREgn52OKCHsqs/edit#gid=1005857707

File: deployments/dev/environment.py [hunk 4]
```diff
@@ -41,26 +41,121 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-24e672db:snapshot/hca_dev_cddab57b68684be4806f395ed9dd635a__20210831_20210907:',
+            'tdr:datarepo-dev-6fdac3db:snapshot/hca_dev_e8808cc84ca0409680f2bba73600cba6__20210902_20210907:',
+            'tdr:datarepo-dev-0c5c20b5:snapshot/hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907:',
+            'tdr:datarepo-dev-e2ab8487:snapshot/hca_dev_03c6fce7789e4e78a27a664d562bb738__20210902_20210907:',
+            'tdr:datarepo-dev-b839d6c7:snapshot/hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20210907:',
+            'tdr:datarepo-dev-311340f6:snapshot/hca_dev_2d8460958a334f3c97d4585bafac13b4__20210902_20210907:',
+            'tdr:datarepo-dev-4c3e6011:snapshot/hca_dev_248fcf0316c64a41b6ccaad4d894ca42__20210907_20210907:',
+            'tdr:datarepo-dev-76de829d:snapshot/hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907:',
+            'tdr:datarepo-dev-24e9529e:snapshot/hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908:',
+            'tdr:datarepo-dev-87beea11:snapshot/hca_dev_ae71be1dddd84feb9bed24c3ddb6e1ad__20210916_20210916:',
```
Comment (amarjandu) [inline deployments/dev/environment.py:154]: here too

**Review by amarjandu — CHANGES_REQUESTED** [review]
Need to update google project id.

**Review by amarjandu — APPROVED** [review]

File: deployments/dev/environment.py [hunk 5]
```diff
@@ -41,26 +41,121 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-e2ab8487:snapshot/hca_dev_03c6fce7789e4e78a27a664d562bb738__20210902_20210907:',
+            'tdr:datarepo-dev-b839d6c7:snapshot/hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20210907:',
+            'tdr:datarepo-dev-311340f6:snapshot/hca_dev_2d8460958a334f3c97d4585bafac13b4__20210902_20210907:',
+            'tdr:datarepo-dev-4c3e6011:snapshot/hca_dev_248fcf0316c64a41b6ccaad4d894ca42__20210907_20210907:',
+            'tdr:datarepo-dev-76de829d:snapshot/hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907:',
+            'tdr:datarepo-dev-24e9529e:snapshot/hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908:',
+            'tdr:datarepo-dev-1dce87e5:snapshot/hca_dev_ae71be1dddd84feb9bed24c3ddb6e1ad__20210916_20210916:',
+            # Managed access snapshots:
+            'tdr:datarepo-dev-d4b988d6:snapshot/hca_dev_a004b1501c364af69bbd070c06dbc17d__20210830_20210903:',
+            'tdr:datarepo-dev-02c59b72:snapshot/hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903:',
```
Comment (hannes-ucsc) [inline deployments/dev/environment.py:157]: I changed my mind. We should list these in the same order as they occur in the spreadsheet.

Each of the two MA sources should have the comment

```
# Managed access:
```

preceding it.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: deployments/dev/environment.py [hunk 6]
```diff
@@ -41,26 +41,122 @@ def env() -> Mapping[str, Optional[str]]:
...
+            'tdr:datarepo-dev-0c5c20b5:snapshot/hca_dev_8c3c290ddfff4553886854ce45f4ba7f__20210902_20210907:',
+            'tdr:datarepo-dev-e2ab8487:snapshot/hca_dev_03c6fce7789e4e78a27a664d562bb738__20210902_20210907:',
+            'tdr:datarepo-dev-b839d6c7:snapshot/hca_dev_e526d91dcf3a44cb80c5fd7676b55a1d__20210902_20210907:',
+            'tdr:datarepo-dev-311340f6:snapshot/hca_dev_2d8460958a334f3c97d4585bafac13b4__20210902_20210907:',
+            'tdr:datarepo-dev-4c3e6011:snapshot/hca_dev_248fcf0316c64a41b6ccaad4d894ca42__20210907_20210907:',
+            'tdr:datarepo-dev-76de829d:snapshot/hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907:',
+            'tdr:datarepo-dev-24e9529e:snapshot/hca_dev_f83165c5e2ea4d15a5cf33f3550bffde__20210901_20210908:',
+            'tdr:datarepo-dev-1dce87e5:snapshot/hca_dev_ae71be1dddd84feb9bed24c3ddb6e1ad__20210916_20210916:',
+            # Managed access:
+            'tdr:datarepo-dev-02c59b72:snapshot/hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903:',
```
Comment (hannes-ucsc) [inline deployments/dev/environment.py:158]: Not consistent with sheet:

image

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Won't review until solution for #3448 lands. Need written assertion that this works in author's personal deployment.

**Comment by jessebrennan** [comment]
When trying to test this in my personal deployment I encountered this error during the Terraform deploy stage for each lambda function:
```
Error: Error modifying Lambda Function Configuration azul-indexer-jesse1-indexercachehealth: RequestEntityTooLargeException: 
        status code: 413, request id: d202cbf6-19de-4ac3-81ee-93bbead087f7

  on indexer/chalice.tf.json line 74, in resource[0].aws_lambda_function[0].indexercachehealth:
  74:                     }
```

The [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/troubleshooting-deployment.html#troubleshooting-deployment-InvalidParameterValueException1) explain:
>The maximum size of the variables object that is stored in the function's configuration must not exceed 4096 bytes

Counting the number of characters in the formatted JSON representation of this object I got 15394 bytes. Stripping out whitespace removes only ~1500 bytes.

Removing the `AZUL_TDR_SOURCES` variable but keeping whitespaces give 4088 bytes. Because of this, I assume that whitespace is automatically stripped out before this is uploaded to AWS.

Nonetheless, we need to figure out a new way to store the environment, or perhaps compress it, if we want to index all of these snapshots.

**Comment by jessebrennan** [comment]
```

Error: Creating CloudWatch Log Group failed: ResourceAlreadyExistsException: The specified log group already exists:  The CloudWatch Log Group '/aws/apigateway/azul-indexer-jesse1' already exists.

  on api_gateway.tf.json line 140, in resource[6].aws_cloudwatch_log_group[0].indexer:
 140:                     }



Error: Creating CloudWatch Log Group failed: ResourceAlreadyExistsException: The specified log group already exists:  The CloudWatch Log Group '/aws/apigateway/azul-service-jesse1' already exists.

  on api_gateway.tf.json line 385, in resource[15].aws_cloudwatch_log_group[0].service:
 385:                     }


make[1]: *** [Makefile:31: apply] Error 1
make[1]: Leaving directory '/home/jesse/gi/azul/terraform'
make: *** [Makefile:81: deploy] Error 2
```
occurred while deploying in personal deployment. Not sure if this is related.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
I'll provide demo instructions later.

**Comment by noah-aviel-dove** [comment]
Failed in GitLab due to inconsistent snapshot storage locations: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/30152

The offending snapshots (`location != 'US'`) are 

```
TDRClient.TDRSource(project='datarepo-dev-bdc9f342', id='d5222430-0f7a-4bd1-8706-c58b4397a9b8', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-12b7a9e1', id='58a0eacf-0caa-449b-9fc8-4fe629e99176', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-02c59b72', id='cddb722a-4229-47bb-ba86-736b7898406f', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-71926fdc', id='403ef1b1-7b7f-4e13-bf4a-d898f047310a', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-dbc582d9', id='79f24905-a9e2-4328-91da-fc18036b77eb', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-10f0610a', id='2c5901f6-67bf-40cc-aea1-621a60c19baa', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-71de019e', id='768c4624-c07b-4496-8b7b-76275a149774', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-78bae095', id='f142a2f7-4c50-458d-a047-27bdb8e0d367', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-a9252919', id='90c2d68f-a000-4c8d-a493-8be93a71b44d', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-1c2c69d9', id='c594fc9a-f1c5-4770-acaa-013c2852e660', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-a198b032', id='38c7e1c4-db48-4cd7-bdf2-36bebc169eac', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-d4b988d6', id='2c4ec571-f489-4277-b215-da8dedd86803', location='us-central1')
TDRClient.TDRSource(project='datarepo-dev-7b7daff7', id='385eb563-005e-43ed-adb7-d3c9e0a8ae93', location='us-central1')
```

**Comment by hannes-ucsc** [comment]
Reindex in `dev` resulted in 16 failed notifications, no failed tallies.

Due to the catalog duplication, there are 8 distinct failures.

7 tracked in https://github.com/HumanCellAtlas/dcp2/issues/45
1 tracked in https://github.com/HumanCellAtlas/dcp2/issues/48

**Comment by hannes-ucsc** [comment]
@jessebrennan, the checklist item "Announce whether reindexing these snapshots succeeded or failed" is lacking the information *where* to announce that.

**Comment by hannes-ucsc** [comment]
I made the announcement: https://humancellatlas.slack.com/archives/C01360XN04S/p1633364377176400?thread_ts=1631840213.076200&cid=C01360XN04S

File: UPGRADING.rst [hunk 7]
```diff
@@ -10,6 +10,15 @@ branch that does not have the listed changes, the steps would need to be
 reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
+
+HumanCellAtlas/dcp2#17 TDR dev dataset is stale
+===============================================
+
+Before upgrading to this commit, run::
+
+    python scripts/reindex.py --delete --catalog dcp2ebi
```
Comment (hannes-ucsc) [inline UPGRADING.rst:19]: This doesn't delete the corresponding IT catalog `it2ebi` and the flag name is `--catalogs` not `--catalog`.

**Comment by hannes-ucsc** [comment]
Another announcement: https://humancellatlas.slack.com/archives/C01360XN04S/p1633367887177300?thread_ts=1631635383.071300&cid=C01360XN04S

**Pull request merged**